src/jit/flowgraph.cpp

```diff
@@ -25694,3 +25694,211 @@ bool Compiler::fgUseThrowHelperBlocks()
 {
     return !opts.compDbgCode;
 }
+
+//------------------------------------------------------------------------
+// fgTailMergeThrows: Tail merge throw blocks and blocks with no return calls.
+//
+// Notes:
+//    Scans the flow graph for throw blocks and blocks with no return calls
+//    that can be merged, and opportunistically merges them.
+//
+//    Does not handle throws yet as the analysis is more costly and less
+//    likely to pay off.
+void Compiler::fgTailMergeThrows()
+{
+    JITDUMP("\n*************** In fgTailMergeThrows\n");
+
+    if (opts.MinOpts())
+    {
+        JITDUMP("Method compiled with minOpts, no merging.\n");
+        return;
+    }
+
+    if (opts.compDbgCode)
+    {
+        JITDUMP("Method compiled with debug codegen, no merging.\n");
+        return;
+    }
+
+    struct ThrowHelperKey
+    {
+        BasicBlock*  m_block;
+        GenTreeCall* m_call;
+
+        ThrowHelperKey() : m_block(nullptr), m_call(nullptr)
+        {
+        }
+        ThrowHelperKey(BasicBlock* block, GenTreeCall* call) : m_block(block), m_call(call)
+        {
+        }
+    };
+
+    struct ThrowHelperKeyFuncs
+    {
+    public:
+        static bool Equals(const ThrowHelperKey& x, const ThrowHelperKey& y)
+        {
+            return BasicBlock::sameEHRegion(x.m_block, y.m_block) && GenTreeCall::Equals(x.m_call, y.m_call);
```

Comment:
Beware that the `GenTree` equality check ignores `gtCallMoreFlags`, most of `gtFlags` and has at least one bug (I ran into it a few days ago - it ignores the size of `GT_BLK` nodes). I'm going to fix the bug soon but it also turns out that `GenTree::Compare` is not widely used in the JIT so it's possible to run into more issues when a new use is added.

Reply:
And if you were perhaps considering to use `gtHashValue` to deal with the more general case - that one also has some issues and it's only used by debug functionality.

Reply:
Good point. Hadn't really looked closely at the compare code.

For the more general case where we try and merge throws (see linked issue) I wasn't relying on tree hashes.